src/Sentry/SentryStackFrame.cs

```diff
@@ -204,19 +205,43 @@ public void ConfigureAppFrame(SentryOptions options)
 
     private void ConfigureAppFrame(SentryOptions options, string parameter, bool mustIncludeSeparator)
     {
-        var resolver = (string prefix) =>
+        var resolver = (string prefixOrPattern) =>
         {
-            if (parameter.StartsWith(prefix, StringComparison.Ordinal))
+            try
             {
-                if (mustIncludeSeparator)
+                var substringOrRegex = new SubstringOrRegexPattern(prefixOrPattern, StringComparison.Ordinal);
+                if (!mustIncludeSeparator)
                 {
-                    return parameter.Length > prefix.Length && parameter[prefix.Length] == '.';
+                    return substringOrRegex.IsMatch(parameter);
                 }
-                return true;
+
+                if (substringOrRegex.Regex is { } regex)
+                {
+                    // Check for any regex match followed by a separator
+                    foreach (Match match in regex.Matches(parameter))
+                    {
+                        if (parameter.Length > match.Value.Length && parameter[match.Value.Length] == '.')
+                        {
+                            return true;
+                        }
+                    }
+                }
+                else
+                {
+                    // Check for a substring followed by a separator
+                    return parameter.Length > prefixOrPattern.Length && parameter[prefixOrPattern.Length] == '.';
+                }
+            }
+            catch (Exception ex)
+            {
+                options.LogDebug("Error matching prefixOrPattern `{0}` with parameter `{1}`: {2}", prefixOrPattern, parameter, ex.Message);
             }
+
             return false;
         };
-        InApp = options.InAppInclude?.Any(resolver) == true || options.InAppExclude?.Any(resolver) != true;
+        var matchesInclude = options.InAppInclude?.Any(resolver) == true;
+        var matchesExclude = () => options.InAppExclude?.Any(resolver) ?? false;
+        InApp = matchesInclude || !matchesExclude();
```

Comment:
Constructing a new SubstringOrRegexPattern (which parses a regex from string) on each stack frame doesn't seem like a right thing to do. How about changing the include/exclude list to already consist of `SubstringOrRegexPattern`s? 
Also, I've never been a fan of creating a lambda here for each frame, especially considering profiling, where each sample has a stack trace that needs to be processed (although just once for each unique frame). How about we get rid of this and make InAppInclude/Exclude a function instead? Taking string parameter and returning bool.

Reply:
We can easily change the lambda to a function 👍🏻 

There's a TODO on the options to change the way these get stored from `List<string>` to `List<SubstringOrRegex>` (which would mean the regex expressions would only be parsed once per program execution) as that would technically be a breaking change to the API. I'll see if I can do something to leave the public API as it is but back it with a `List<SubstringOrRegex>`.

Reply:
Oh... in fact I see [both of those properties are internal](https://github.com/getsentry/sentry-dotnet/blob/b74a161f6b7821c6fe45dd114c8ae6dc3a76ec9a/src/Sentry/SentryOptions.cs#L237-L264) 🎉

Reply:
I ended up having to create a new class for this as we actually wanted `PrefixOrRegexPattern` not `SubstringOrRegexPattern`. 

I've done away with the lambdas as well (these are now inline functions).

Reply:
> I've done away with the lambdas as well (these are now inline functions).
> 
> The tests seem to pass now. 🎈

So indeed it seems like it has been the performance issue. Maybe we should check check that we have some benchmarks for these critical paths and introduce https://github.com/marketplace/actions/continuous-benchmark to the repo.

Reply:
BTW there already is a benchmark so we would have probably caught this if we had the action in place:
https://github.com/getsentry/sentry-dotnet/blob/24c159fac119315e21df6138dc04bfe3ccd573c2/benchmarks/Sentry.Benchmarks/StackFrameBenchmarks.cs#L40

Reply:
> BTW there already is a benchmark so we would have probably caught this if we had the action in place:

Interesting. And how would that typically integrate with CI? Is it a bit like code coverage where you're comparing to previous runs or something?

Reply:
For info on benchmarking options, see:
- https://github.com/getsentry/sentry-dotnet/issues/3332

Reply:
> And how would that typically integrate with CI? Is it a bit like code coverage where you're comparing to previous runs or something?

Yes, I'd expect a comment on the PR with previous and current results. I haven't worked with the action myself though so maybe it doesn't have that (yet)